src/main.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+ * Display.c
+ *
+ *  Created on: 14.08.2017
+ *      Author: darek
+ */
+// #include <driver/i2c.h>
+#define LOG_LOCAL_LEVEL ESP_LOG_VERBOSE
+#include "esp_log.h"
+#include <freertos/FreeRTOS.h>
+#include <freertos/task.h>
+#include <esp_system.h>
+#include "imu.h"
+#include "sdkconfig.h"
+#include <cstring>
+#include <array>
+#include <string.h>
+#include "driver/uart.h"
+#include "driver/ledc.h"
+#include "driver/gpio.h"
+
+
+const int DELAY_MS = 50;
+
+// uart communication 
+#define TXD_PIN (GPIO_NUM_17)
+#define RXD_PIN (GPIO_NUM_16)
+#define TX_BUFFER_SIZE 16 // sending packet
+static const int RX_BUF_SIZE = 1024;
+
+// MOTOR CONSTANTS
+const float NO_LOAD_SPEED = 7.50f;
+const float STALL_TORQUE = 1.05f;
+#define LEDC_TIMER              LEDC_TIMER_0
+#define LEDC_MODE               LEDC_LOW_SPEED_MODE
+#define LEDC_CHANNEL            LEDC_CHANNEL_0
+#define DUTY_RES                LEDC_TIMER_8_BIT// Set duty resolution to 13 bits
+#define LEDC_FREQUENCY          (50000) // Frequency in Hertz. Set frequency at 4 kHz
+
+
+struct Motor {
+  gpio_num_t encoder;  // yellow
+  gpio_num_t pwm;// blue
+  gpio_num_t dir; // white
+};
+
+const Motor motor1 = { GPIO_NUM_18, GPIO_NUM_5, GPIO_NUM_19 };   // left
+const Motor motor2 = { GPIO_NUM_12, GPIO_NUM_14, GPIO_NUM_27 };   // right
+
+
+static IMUData imu_data;
+
+
+extern "C" {
+  void app_main(void);
+}
+
+void init_uart(void);
+void init_motors(void);
+void imu_cb(IMUData data) { imu_data = data; }
+static void tx_task(void*); // read imu_data and send it over uart
+static void rx_task(void*); // receive motor_cmd and set torque of the motors
+void set_torque(float torque, const Motor& motor);
+
+
+void app_main() {
+  init_uart();
+  init_motors();
+  // xTaskCreate(task_initI2C, "task_initI2C", 2048, NULL, 5, NULL);
+  xTaskCreate(tx_task, "tx_task", 4096, NULL, 5, NULL);
+  xTaskCreate(rx_task, "rx_task", 4096, NULL, 5, NULL);
+}
+
+static void rx_task(void* arg) {
+  static const char* RX_TASK_TAG = "RX_TASK";
+  esp_log_level_set(RX_TASK_TAG, ESP_LOG_INFO);
+  char read_buf[RX_BUF_SIZE];
+  float motor_cmd_1;
+  float motor_cmd_2;
+  while (1) {
+    // clear buffer
+    for (size_t i = 0; i < RX_BUF_SIZE; i++) {
+      read_buf[i] = 0x00;
+    }
+    const int rxBytes = uart_read_bytes(UART_NUM_1, read_buf, RX_BUF_SIZE, DELAY_MS / portTICK_PERIOD_MS);
+
+    if (rxBytes <= 0) {
+      ESP_LOGI(RX_TASK_TAG, "No bytes, received\n");
+      continue;
+    }
+    ESP_LOGI(RX_TASK_TAG, "Motor cmd : %f\t%f", motor_cmd_1, motor_cmd_2);
+    // ESP_LOGI(RX_TASK_TAG, "Read %d bytes: '%s'", rxBytes, read_buf);
+
+
+    memcpy(&motor_cmd_1, read_buf + 3, sizeof(float));
+    memcpy(&motor_cmd_2, read_buf + 7, sizeof(float));
```

Comment:
Could you leave a comment to explain where these magic numbers come from?

Reply:
Will do!